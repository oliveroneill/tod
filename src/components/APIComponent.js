import React, { Component } from 'react';
import {
  View,
  Alert
} from 'react-native';
import PropTypes from 'prop-types';

import LoadingScreen from '../components/LoadingScreen.js';
import TodAPI from '../utils/TodAPI.js';

class APIComponent extends Component {
  state={
    errored: false,
    loading: true
  };
  constructor(props) {
    super(props);
    // bind functions
    this.initialise = this.initialise.bind(this);
    this.setupCurrentLocation = this.setupCurrentLocation.bind(this);
  }

  initialise() {
    this.setState({'errored': false});
    this.setState({'loading': true});
    let { api, onNotification } = this.props;
    api.setup(
      this.setupCurrentLocation,
      onNotification,
      function(){
        this.setState({'errored': true})
      }.bind(this)
    );
  }

  setupCurrentLocation() {
    let { onComplete } = this.props;
    navigator.geolocation.getCurrentPosition(
      function(origin) {
        let lat = origin.coords.latitude;
        let lng = origin.coords.longitude;
        this.setState({'loading': false});
        onComplete(lat, lng);
      }.bind(this),
      function(error){
        Alert.alert("Please enable location and try again");
        this.setState({'errored': true})
      }.bind(this),
      {enableHighAccuracy: false}
    );
  }

  componentWillMount() {
    this.initialise();
  }

  render() {
    let {children} = this.props;
    if (this.state.loading) {
      return (
        <LoadingScreen
          errored={this.state.errored}
          loadingMessage="Logging in..."
          errorMessage="Unfortunately we couldn't log you in. Please ensure you have an internet connection."
          retry={this.initialise}
        />
      )
    }
    return children;
  }
}

APIComponent.propTypes = {
  api: PropTypes.instanceOf(TodAPI).isRequired,
  onNotification: PropTypes.func.isRequired,
  onComplete: PropTypes.func.isRequired,
}

export default APIComponent;