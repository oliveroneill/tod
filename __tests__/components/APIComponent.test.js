import React from 'react';
import {
  Text
} from 'react-native';
import renderer from 'react-test-renderer';

import APIComponent from '../../src/components/APIComponent.js';
import TodAPI from '../../src/utils/TodAPI.js';

function mockTodAPI() {
  // Mock out the interface to TodAPI
  // So that we can avoid warning with PropTypes
  TodAPI.prototype.setup = jest.fn();
  TodAPI.prototype.sendToken = jest.fn();
  TodAPI.prototype.scheduleTrip = jest.fn();
  TodAPI.prototype.getScheduledTrips = jest.fn();
  TodAPI.prototype.getRoutes = jest.fn();
  TodAPI.prototype.enableDisableTrip = jest.fn();
  TodAPI.prototype.deleteTrip = jest.fn();
}

function setup() {
  navigator.geolocation.getCurrentPosition = jest.fn();
  mockTodAPI();
  let api = new TodAPI();
  return api;
}

describe('APIComponent', () => {
  beforeEach(() => {
    navigator = {
      geolocation: {
        getCurrentPosition: jest.fn()
      }
    };
  })

  it('shows spinner while loading', () => {
    let api = setup();
    let component = renderer.create(
      <APIComponent
        api={api}
        onLocation={jest.fn()}
        onNotification={jest.fn()}
      />
    );
    expect(navigator.geolocation.getCurrentPosition).not.toHaveBeenCalled();
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('checks location after logging in', () => {
    let api = setup();
    api.setup = jest.fn().mockImplementation((onRegister, onNotification, onError) => {
      onRegister();
    });
    let component = renderer.create(
      <APIComponent
        api={api}
        onLocation={jest.fn()}
        onNotification={jest.fn()}
      />
    );
    expect(navigator.geolocation.getCurrentPosition).toHaveBeenCalled();
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('shows children when location and logging in complete', () => {
    let api = new TodAPI();
    api.setup = jest.fn().mockImplementation((onRegister, onNotification, onError) => {
      onRegister();
    });
    let onComplete = jest.fn();
    let mockLocFunction = jest.fn().mockImplementation((onLocation, onError) => {
      onLocation({coords: {latitude: 1, longitude: 1}});
    });
    navigator.geolocation.getCurrentPosition = mockLocFunction;
    let component = renderer.create(
      <APIComponent
        api={api}
        onLocation={onComplete}
        onNotification={jest.fn()}
      />
    );
    expect(mockLocFunction).toHaveBeenCalled();
    expect(onComplete).toHaveBeenCalled();
  });

  it('shows error case on send token failure', () => {
    let api = setup();
    api.setup = jest.fn().mockImplementation((onRegister, onNotification, onError) => {
      onError();
    });
    let component = renderer.create(
      <APIComponent
        api={api}
        onLocation={jest.fn()}
        onNotification={jest.fn()}
      />
    );
    expect(navigator.geolocation.getCurrentPosition).not.toHaveBeenCalled();
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('shows error case when location is disabled', () => {
    let api = new TodAPI();
    api.setup = jest.fn().mockImplementation((onRegister, onNotification, onError) => {
      onRegister();
    });
    let onComplete = jest.fn();
    let mockLocFunction = jest.fn().mockImplementation((onLocation, onError) => {
      onError();
    });
    navigator.geolocation.getCurrentPosition = mockLocFunction;
    let component = renderer.create(
      <APIComponent
        api={api}
        onLocation={onComplete}
        onNotification={jest.fn()}
      />
    );
    expect(mockLocFunction).toHaveBeenCalled();
    expect(onComplete).not.toHaveBeenCalled();
    expect(component.toJSON()).toMatchSnapshot();
  });
});