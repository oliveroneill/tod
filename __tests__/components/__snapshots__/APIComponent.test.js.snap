// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APIComponent checks location after logging in 1`] = `
<View>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="large"
    style={
      Object {
        "height": 80,
      }
    }
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      Logging in...
    </Text>
  </View>
</View>
`;

exports[`APIComponent shows children when location and logging in complete 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  Example child component
</Text>
`;

exports[`APIComponent shows error case on send token failure 1`] = `
<View>
  <View
    style={
      Array [
        Object {
          "marginTop": 80,
        },
        Object {
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingLeft": 10,
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      Unfortunately we couldn't log you in. Please ensure you have an internet connection.
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      accessibilityComponentType="button"
      accessibilityLabel={undefined}
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={undefined}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Try again
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`APIComponent shows error case when location is disabled 1`] = `
<View>
  <View
    style={
      Array [
        Object {
          "marginTop": 80,
        },
        Object {
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingLeft": 10,
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      Unfortunately we couldn't log you in. Please ensure you have an internet connection.
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      accessibilityComponentType="button"
      accessibilityLabel={undefined}
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={undefined}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Try again
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`APIComponent shows spinner while loading 1`] = `
<View>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="large"
    style={
      Object {
        "height": 80,
      }
    }
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      Logging in...
    </Text>
  </View>
</View>
`;
